cmd/flat-network-controller/app/controller.go

```diff
@@ -0,0 +1,247 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"strings"
+	"time"
+)
+
+// Controller represents the flat network controller
+type Controller struct {
+	cacheSynced  cache.InformerSynced
+	store        arktosv1.NetworkLister
+	queue        workqueue.RateLimitingInterface
+	netClientset *arktos.Clientset
+	svcClientset *kubernetes.Clientset
+	recorder     record.EventRecorder
+}
+
+// New creates the controller object
+func New(netClientset *arktos.Clientset, svcClientset *kubernetes.Clientset, informer arktosinformer.NetworkInformer) *Controller {
+	utilruntime.Must(arktoscheme.AddToScheme(scheme.Scheme))
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: svcClientset.CoreV1().EventsWithMultiTenancy(metav1.NamespaceAll, metav1.TenantAll)})
+
+	return &Controller{
+		store:        informer.Lister(),
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		cacheSynced:  informer.Informer().HasSynced,
+		netClientset: netClientset,
+		svcClientset: svcClientset,
+		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "flat-network-controller"}),
+	}
+}
+
+// Run starts the control loop with workers processing the items
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("starting flat network controller")
+	klog.V(5).Info("waiting for informer caches to sync")
+	if !cache.WaitForCacheSync(stopCh, c.cacheSynced) {
+		klog.Error("failed to wait for cache to sync")
+		return
+	}
+
+	klog.V(5).Info("staring workers of flat network controller")
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.V(5).Infof("%d workers started", workers)
+	<-stopCh
+	klog.Info("shutting down flat network controller")
+}
+
+func (c *Controller) runWorker() {
+	for {
+		item, shutdown := c.queue.Get()
+		if shutdown {
+			break
+		}
+
+		c.process(item)
+	}
+}
+
+// processItem will read a single work item off the work queue and attempt to process it
+func (c *Controller) process(item interface{}) {
+	defer c.queue.Done(item)
+
+	key, ok := item.(string)
+	if !ok {
+		klog.Errorf("unexpected item in queue: %v", item)
+		c.queue.Forget(item)
+		return
+	}
+
+	tenant, name, err := parseKey(key)
+	if err != nil {
+		klog.Errorf("unexpected string in queue; discarding: %s", key)
+		c.queue.Forget(item)
+		return
+	}
+
+	net, err := c.store.NetworksWithMultiTenancy(tenant).Get(name)
+	if err != nil {
+		klog.Warningf("failed to retrieve network in local cache by tenant %s, name %s: %v", tenant, name, err)
+		c.queue.Forget(item)
+		return
+	}
+
+	if net.Spec.Type != "flat" {
+		klog.V(5).Infof("network %s/%s is of type %q; ignored", net.Tenant, net.Name, net.Spec.Type)
+		c.queue.Forget(item)
```

Comment:
I found every path in this func, either success or fail, will call c.queue.Forget(item). If we need to forget this item anyway, consider call it in the func of runWorker(), which the caller of process().

Reply:
not all branches call queue.Forget()